File: BrainExtraction/T2/trainModel.R
library( ANTsR )
library( ANTsRNet )
library( keras )
library( tensorflow )
library( ggplot2 )

keras::backend()$clear_session()

Sys.setenv( "CUDA_VISIBLE_DEVICES" = "1" )

baseDirectory <- '/home/ntustison/Data/BrainExtractionFlairT2/'
scriptsDirectory <- paste0( baseDirectory, 'ScriptsT2/' )
source( paste0( scriptsDirectory, 'batchGenerator.R' ) )

templateDirectory <- paste0( baseDirectory, '../Templates/Kirby/SymmetricTemplate/' )
template <- antsImageRead( paste0( templateDirectory, "S_template3_resampled2.nii.gz" ) ) %>% iMath( "Normalize" )
templateSize <- dim( template )

################################################
#
#  Create the model and load weights
#
################################################

classes <- c( "background", "brain" )
numberOfClassificationLabels <- length( classes )
imageModalities <- c( "T2" )
channelSize <- length( imageModalities )

unetModel <- createUnetModel3D( c( templateSize, channelSize ),
  numberOfOutputs = numberOfClassificationLabels, mode = 'classification',
  numberOfLayers = 4, numberOfFiltersAtBaseLayer = 8, dropoutRate = 0.0,
  convolutionKernelSize = c( 3, 3, 3 ), deconvolutionKernelSize = c( 2, 2, 2 ),
  weightDecay = 1e-5 )

brainWeightsFileName <- paste0( scriptsDirectory, "/brainExtractionT2Weights.h5" )
if( file.exists( brainWeightsFileName ) )
  {
  load_model_weights_hdf5( unetModel, filepath = brainWeightsFileName )
  # } else {
  # stop( "Weights file doesn't exist.\n" )  
  }
unetModel %>% compile(
  optimizer = optimizer_adam(),
  loss = "categorical_crossentropy",
  metrics = 'accuracy' )

################################################
#
#  Load the brain data
#
################################################

cat( "Loading brain data.\n" )

brainBaseDataDirectory <- '/home/ntustison/Data/BrainExtractionFlairT2/Data/'
brainImages <- list.files( path = brainBaseDataDirectory, pattern = "*t2.nii.gz", recursive = TRUE, full.names = TRUE )

trainingImageFiles <- c()
trainingBrainMaskFiles <- c()

pb <- txtProgressBar( min = 0, max = length( brainImages ), style = 3 )
for( i in seq_len( length( brainImages ) ) )
  {
  setTxtProgressBar( pb, i )

  image <- brainImages[i]
  mask <- gsub( "-t2", "-mask", image )

  if( ! file.exists( mask ) )
    {
    next  
    # stop( "Mask doesn't exist." )  
    }

  # # Do a quick check  
  # maskImage <- antsImageRead( mask )
  # if( sum( maskImage ) < 1000 )
  #   {
  #   cat( "mask: ", mask, "\n" )  
  #   next  
  #   }
  # imageArray <- as.array( antsImageRead( image ) )
  # if( any( is.na( imageArray ) ) || ( sum( imageArray ) <= 1000 ) )
  #   {
  #   cat( "image: ", image, "\n" )  
  #   next  
  #   }
    
  trainingImageFiles <- append( trainingImageFiles, image )
  trainingBrainMaskFiles <- append( trainingBrainMaskFiles, mask )
  }
cat( "\n" )  

cat( "Total training image files: ", length( trainingImageFiles ), "\n" )

cat( "\nTraining\n\n" )

###
#
# Set up the training generator
#

batchSize <- 8L

# Split trainingData into "training" and "validation" componets for
# training the model.

numberOfData <- length( trainingImageFiles )
sampleIndices <- sample( numberOfData )

validationSplit <- floor( 0.8 * numberOfData )
trainingIndices <- sampleIndices[1:validationSplit]
numberOfTrainingData <- length( trainingIndices )
validationIndices <- sampleIndices[( validationSplit + 1 ):numberOfData]
numberOfValidationData <- length( validationIndices )

###
#
# Run training
#

track <- unetModel %>% fit_generator(
  generator = batchGenerator( batchSize = batchSize,
                              imageSize = templateSize, 
                              template = template,
                              images = trainingImageFiles[trainingIndices],
                              brainMasks = trainingBrainMaskFiles[trainingIndices],
                              doRandomContralateralFlips = FALSE,
                              doDataAugmentation = FALSE
                            ),
  steps_per_epoch = 100L,
  epochs = 200L,
  validation_data = batchGenerator( batchSize = batchSize,
                                    imageSize = templateSize, 
                                    template = template,
                                    images = trainingImageFiles[validationIndices],
                                    brainMasks = trainingBrainMaskFiles[validationIndices],
                                    doRandomContralateralFlips = FALSE,
                                    doDataAugmentation = FALSE
                                  ),  
  validation_steps = 40L,
  callbacks = list(
    callback_model_checkpoint( brainWeightsFileName,
      monitor = 'loss', save_best_only = TRUE, save_weights_only = TRUE,
      verbose = 1, mode = 'auto' ), 
     callback_reduce_lr_on_plateau( monitor = 'loss', factor = 0.3,
       verbose = 1, patience = 10, mode = 'auto' ),
    callback_early_stopping( monitor = 'loss', min_delta = 0.001,
      patience = 20 )
  )
)

save_model_weights_hdf5( unetModel, filepath = brainWeightsFileName )

File: NVP/train_model.py
import os
import glob

os.environ["CUDA_VISIBLE_DEVICES"] = "0"
os.environ["TF_USE_LEGACY_KERAS"] = "True"

import tensorflow as tf

from batch_generator import batch_generator
from create_normalizing_flow_model import create_normalizing_flow_model

base_directory = '/home/ntustison/Data/NVP/'
scripts_directory = base_directory + 'Scripts/'

data_directory = '/home/ntustison/Data/Inpainting/AlignedToNKISlices/'
brain_slice_files = glob.glob(data_directory + "*/Images/axial/*.nii.gz")

print("Total number of training brain slices: ", str(len(brain_slice_files)))

################################################
#
#  Create the model and load weights
#
################################################

# Copy from https://github.com/aganse/flow_models/blob/main/train.py

image_size = (256, 256, 1)

nvp_model = create_normalizing_flow_model(image_size, 
    hidden_layers=[512, 512], flow_steps=6, regularization=0.01,
    validate_args=False)
nvp_model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=2e-4),
                  loss='negative_log_likelihood')
nvp_model.summary()

weights_filename = scripts_directory + "nvp_t1_axial.weights.h5"

if os.path.exists(weights_filename):
    print("Loading " + weights_filename)
    nvp_model.load_weights(weights_filename)

###
#
# Set up the training generator
#

batch_size = 128

generator = batch_generator(batch_size=batch_size,
                            input_image_files=brain_slice_files,
                            image_size=image_size,
                            do_histogram_intensity_warping=True,
                            do_simulate_bias_field=True,
                            do_add_noise=True,
                            do_random_transformation=True,
                            do_random_contralateral_flips=True,
                            do_resampling=False,
                            verbose=False)

track = nvp_model.fit(x=generator, epochs=100, verbose=1, steps_per_epoch=64,
    callbacks=[
       tf.keras.callbacks.ModelCheckpoint(weights_filename, monitor='negative_log_likelihood',
           save_best_only=True, save_weights_only=True, mode='auto', verbose=1),
       tf.keras.callbacks.ReduceLROnPlateau(monitor='negative_log_likelihood', factor=0.95,
          verbose=1, patience=20, mode='auto'),
    #    keras.callbacks.EarlyStopping(monitor='loss', min_delta=0.000001,
    #       patience=20)
       ]
   )

nvp_model.save_weights(weights_filename)


